flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointCustomRescaleITCase.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.test.checkpointing;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.api.common.JobStatus;
+import org.apache.flink.api.common.functions.Partitioner;
+import org.apache.flink.configuration.CheckpointingOptions;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobgraph.SavepointRestoreSettings;
+import org.apache.flink.runtime.minicluster.MiniCluster;
+import org.apache.flink.runtime.minicluster.MiniClusterConfiguration;
+import org.apache.flink.streaming.api.environment.CheckpointConfig;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.sink.SinkFunction;
+import org.apache.flink.util.FileUtils;
+
+import org.junit.After;
+import org.junit.Ignore;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+
+import static org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.createLocalEnvironment;
+import static org.junit.Assert.fail;
+
+/** Integration test for performing rescale of unaligned checkpoint with custom partitioner. */
+public class UnalignedCheckpointCustomRescaleITCase {
```

Comment:
Actually, instead of creating custom test, can you create a new `Topology` in the pre-existing `UnalignedCheckpointRescaleITCase`?

Reply:
I though about it, but in `UnalignedCheckpointRescaleITCase` used `Long` as stream data. But for the regular repeatability of the test it was necessary to use the `String`

Reply:
>  But for the regular repeatability of the test it was necessary to use the String.

There are plenty of solutions to this problem:

1. You are encoding in your `String` record three fields: `partition`, `index` and some `payload`. But I don't see you using anything but the `partition`, so you could convert your whole `String` record into `Long` record with value just for the `partition`.
2. You can encode any number of fields in single `Long` just as well. The easiest would be sth like that:
```
long encode(int partition, int index, int payload) {
  checkArgument(partition < 1024);
  checkArgument(index < 1024);
  checkArgument(payload < Long.MAX_VALUE / (1024*1024));
  return partition + index * 1024 + payload * 1024 * 1024;
}
``` 
3. You could always most likely change the record type for `UnalignedCheckpointRescaleITCase`. `String` there should work just as fine, but that's probably more work vs option 1.

> but I'm not sure. Also in UnalignedCheckpointRescaleITCase rescale full graph (change parallelism for all vertexes), but need to change only one vertex

When creating job graph in `Topology`  (`UnalignedCheckpointRescaleITCase.Topology#create`), you can set parallelism per vertex and AFAIK that will override the `env.setParallelism(...)` setting. So for the downstream vertex/task, that you don't want to change parallelism, you can call:
```
                    .addSink(new StringSink(createCheckpoint ? 100 : 1000))
                    .name("sink")
                    .setParallelism(3);
```
While keep the upstream source's parallelism controlled by the environment.

Reply:
An important point for test repeatability is to split the record into two parts (one at the sink input, one at the source output). That's why we use a string with a sufficiently long length and all bytes set to 1 to be sure to read data leading to exception in case of splitting. 
That's why need to realize 3rd option

Reply:
> You could always most likely change the record type for UnalignedCheckpointRescaleITCase. String there should work just as fine, but that's probably more work vs option 1.

I've just realised that I don't see any place in the `UnalignedCheckpointRescaleITCase` that hardcodes the type of the record. You can add a new `Topology` (`UnalignedCheckpointRescaleITCase.Topology`) that creates any JobGraph, so you can keep your proposed records format. Also I wouldn't mind if you re-used the same `LongSource`, but changed the record type from `Long` to some POJO of `Long` and `String payload`
```
public static class Record {
  private final long value;
  private final @Nullable String payload;
  (...)
}
```
where `payload` length/size would be configurable, between 0 (`payload == null`) to whatever value you would configured (`3713`?).

This way you won't be duplicating all of the setup code and you will leverage the same

Reply:
A difficult way out of the situation, but quite feasible, ok.  Thank you

Reply:
I'm sorry, but I think this is too complicated a way to implement the test. That's why I came up with an idea that is not as elegant and correct as yours, but is easier to implement and allows you to do without changing less code. Please check it out. Thank you for your patience